infrastructure/index.ts

```diff
@@ -257,47 +242,121 @@ function getMimeType(filePath: string): string | undefined {
-                },
-                {
-                    parent: contentBucket,
-                });
-        } else {
-            const s3Redirect = new aws.s3.BucketObject(
-                relativeFilePath,
-                {
-                    ...baseArgs,
-                    source: new pulumi.asset.FileAsset("/dev/null"), // Empty file.
-                    websiteRedirect: translateRedirect(filePath, redirect),
-                },
-                {
-                    parent: contentBucket,
-                });
-        }
+// Certain files, like JavaScript and CSS files produced during the build, must exist
+// before the website's HTML files can use them (otherwise, pages won't render properly,
+// or may be unusable), and others, like fonts and images, would be nice to have there in
+// advance, too. Extensions listed here are used for identifying files that should exist
+// on S3 before any HTML pages are updated to reference them.
+const assetTypes = [
+    "js",
+    "css",
+    "png",
+    "jpg",
+    "gif",
+    "svg",
+    "pdf",
+    "eot",
+    "ttf",
+    "woff",
+    "woff2",
+];
+
+// Keep track of the asset paths we encounter and the resources we create from them, so we
+// can declare the necessary dependency relationships.
+const assetPaths: string[] = [];
+const assetResources: aws.s3.BucketObject[] = [];
+
+// Find all files matching the patterns above, creating a new BucketObject for each one.
+// These objects are declared first in order to pass them conditionally as dependencies
+// below.
+assetTypes.forEach(pattern => {
+    glob.sync(`${webContentsRootPath}/**/*.${pattern}`).forEach(assetPath => {
+
+        // Remember the asset path, so we can ignore it later.
+        assetPaths.push(assetPath);
+
+        // Turn the full (local) path into a relative path, so we can use it for an S3 object key.
+        const relativeAssetPath = assetPath.replace(webContentsRootPath + "/", "");
+
+        // Create a resource object with the asset.
+        const assetResource = new aws.s3.BucketObject(
+            relativeAssetPath,
+            {
+                acl: "public-read",
+                key: relativeAssetPath,
+                source: new pulumi.asset.FileAsset(assetPath),
+                contentType: getMimeType(assetPath) || undefined,
+                bucket: contentBucket,
+            },
+            {
+                parent: contentBucket,
+            },
+        );
+
+        // Add the bucket resource to the list of asset dependencies.
+        assetResources.push(assetResource);
     });
+});
+
+// Now create bucket objects for everything else, bypassing anything created already.
+glob.sync(`${webContentsRootPath}/**/*`).forEach(filePath => {
+
+    // Ignore directories.
+    if (fs.lstatSync(filePath).isDirectory()) {
+        return;
+    }
+
+    // Ignore previously encountered asset paths, since they'll already have been declared.
+    if (assetPaths.includes(filePath)) {
```

Comment:
I don't know if it'd make a measurable improvement, but I wonder if it'd be more performant to store these in a `Set<string>` for speeder lookups, instead of iterating over the (potentially entire) array doing string comparisons for each file.

```typescript
const assetPaths = new Set<string>();
```

```typescript
// Remember the asset path, so we can ignore it later.
assetPaths.add(assetPath);
```

```typescript
// Ignore previously encountered asset paths, since they'll already have been declared.
if (assetPaths.has(filePath)) {
    return;
}
```

Reply:
Ah, good suggestion!